core/src/main/java/org/jdbi/v3/core/cache/internal/InternalJdbiCache.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.cache.internal;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+
+import org.jdbi.v3.core.cache.JdbiCache;
+import org.jdbi.v3.core.cache.JdbiCacheBuilder;
+import org.jdbi.v3.core.cache.JdbiCacheLoader;
+
+public final class InternalJdbiCache<K, V> implements JdbiCache<K, V> {
+
+    private final ConcurrentMap<K, V> cache;
+    private final ConcurrentLinkedQueue<K> expungeQueue;
+    private final JdbiCacheLoader<K, V> cacheLoader;
+
+    private final int maxSize;
+
+    private InternalJdbiCache(Builder builder, JdbiCacheLoader<K, V> cacheLoader) {
+        this.cache = new ConcurrentHashMap<>();
+        this.expungeQueue = new ConcurrentLinkedQueue<>();
+        this.cacheLoader = cacheLoader;
+
+        this.maxSize = builder.maxSize;
+    }
+
+    @Override
+    public V get(K key) {
+        try {
+            if (cacheLoader == null) {
+                return cache.get(key);
+            } else {
+                return cache.computeIfAbsent(key, compute(cacheLoader)::create);
+            }
+        } finally {
+            expunge();
+        }
+    }
+
+    @Override
+    public V getWithLoader(K key, JdbiCacheLoader<K, V> cacheLoader) {
+        try {
+            return cache.computeIfAbsent(key, compute(cacheLoader)::create);
+        } finally {
+            expunge();
+        }
+    }
+
+    @Override
+    public CacheStats getStats() {
+        return new CacheStats(cache.size(), maxSize);
+    }
+
+    private JdbiCacheLoader<K, V> compute(JdbiCacheLoader<K, V> delegate) {
+        if (delegate == null) {
```

Comment:
This `null` cascading is slightly concerning. This is used like `compute(cacheLoader)::create` which will not be so happy with `null`. Should we assert that it is not null here, instead?

Reply:
I can rework the logic so that we use a dummy loader that always returns null if no loader was passed in.

Reply:
done. Instead of null, this now returns  `k -> null` which is a lambda for a JdbiCacheLoader that returns null for all keys.